First cut at converting common-custom-user-data-groovy-script to Gradle plugin

bigdaz opened the pull request.

etiennestuder commented on common-custom-user-data-gradle-plugin/build.gradle:
```diff
@@ -0,0 +1,25 @@
+group = "com.gradle"
+version = "0.1"
+
+repositories {
```
For the group and artifact, I suggest to use what we use for Maven but exchange 'Gradle':

```
<dependency>
  <groupId>com.gradle</groupId>
  <artifactId>common-custom-user-data-gradle-plugin</artifactId>
  <version>1.0</version>
</dependency>
```

bigdaz replied:
This is already the case, except for the temporary version `0.1`

etiennestuder commented on common-custom-user-data-gradle-plugin/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.7-bin.zip
```
6.8 is out :-)

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/BuildScanEnhancer.java:
```diff
@@ -0,0 +1,442 @@
+        hexString.append("...");
+        return hexString.toString();
+    }
+}
```
I suggest to copy/paste the code from [here](https://github.com/gradle/gradle-enterprise-build-config-samples/blob/master/common-custom-user-data-maven-extension/src/main/java/com/gradle/CustomBuildScanConfig.java#L28).

etiennestuder replied:
Except the custom data we add only for Gradle.

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/BuildScanEnhancer.java:
```diff
@@ -0,0 +1,442 @@
+        this.buildScan = buildScanExtension;
+        this.rootProject = rootProject;
+        try {
+            messageDigest = MessageDigest.getInstance("SHA-256");
```
I think this could be a static field.

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CommonCustomUserDataGradlePlugin.java:
```diff
@@ -0,0 +1,52 @@
+        });
+    }
+
+    private void applySettingsPlugin(Settings settings) {
```
I wouls swap the order of this and the previous method.

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CommonCustomUserDataGradlePlugin.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private void applyProjectPlugin(Project project) {
+        project.getPlugins().withType(BuildScanPlugin.class, __ -> {
```
What does __ do?

bigdaz replied:
This is a Kotlin-esque way to define a variable that is unused. I'm not sure if there's an idiomatic way to declare an unused variable for a Java lambda.

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CommonCustomUserDataGradlePlugin.java:
```diff
@@ -0,0 +1,52 @@
+
+    private void applySettingsPlugin(Settings settings) {
+        settings.getPlugins().withType(GradleEnterprisePlugin.class, __ ->
+                settings.getGradle().projectsLoaded(gradle -> {
```
Not sure, do we have to wait for the projects being loaded?

bigdaz replied:
This was required to allow us to get project properties, which are used to detect Android Studio invocation. I'll see if there's another way...

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CommonCustomUserDataGradlePlugin.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private void enhanceBuildScan(BuildScanExtension buildScanExtension, Project rootProject) {
+        enhanceBuildScan(new BuildScanEnhancer(buildScanExtension, rootProject));
```
Why do we need to pass the extension to the enhancer?

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CommonCustomUserDataGradlePlugin.java:
```diff
@@ -0,0 +1,52 @@
+        buildScanEnhancer.addGitMetadata();
+        buildScanEnhancer.addTestParallelization();
+        buildScanEnhancer.addTestSystemProperties();
+    }
```
Maybe a matter of taste, but I would probably move these invocations to the enhancer to keep it closer to what we do for the Maven extension.

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CommonCustomUserDataGradlePlugin.java:
```diff
@@ -0,0 +1,52 @@
+        buildScanEnhancer.addTestParallelization();
+        buildScanEnhancer.addTestSystemProperties();
+    }
+}
```
Can you please add a little readme file, too, like done for the Maven extension?
https://github.com/gradle/gradle-enterprise-build-config-samples/tree/master/common-custom-user-data-maven-extension

etiennestuder replied:
Plus a changelog for initial release:
https://github.com/gradle/gradle-enterprise-build-config-samples/blob/master/common-custom-user-data-maven-extension/CHANGELOG.md

etiennestuder commented on common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CommonCustomUserDataGradlePlugin.java:
```diff
@@ -0,0 +1,52 @@
+                })
+        );
+    }
+
```
Can you also add a hook (with empty implementation) for customizing the build cache configuration, like we did it for Maven?
https://github.com/gradle/gradle-enterprise-build-config-samples/blob/master/common-custom-user-data-maven-extension/src/main/java/com/gradle/CustomBuildCacheConfig.java

bigdaz commented:
@etiennestuder I took another stab at this using the Maven Extension sources as a basis, with a goal of minimizing the differences between the Gradle and Maven implementations.
I'm going to close this PR and open a new one. I've addressed all of your feedback in the new PR.
